### PR TITLE
feat (api): poller refacto

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/application/application.go [hunk 1]
```diff
@@ -368,6 +368,20 @@ func DeleteApplication(db gorp.SqlExecutor, applicationID int64) error {
 		return err
 	}
 
+	// Delete poller execution
+	query = `DELETE FROM poller_execution WHERE application_id = $1`
```
Comment (sguiheux) [inline engine/api/application/application.go]: should be in poller package

File: engine/api/poller/cleaner.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	}
+	defer tx.Rollback()
+
+	//Starting with exclusive lock on the table
+	if err := LockPollerExecutions(tx); err != nil {
+		log.Debug("poller.CleanerRun> Unable to take lock : %s", err)
+		return nil, err
+	}
+
+	//Load schedulers
```
Comment (sguiheux) [inline engine/api/poller/cleaner.go]: Load pollers?

File: engine/api/poller/cleaner.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	ps, err := LoadAll(tx)
+	if err != nil {
+		log.Warning("poller.CleanerRun> Unable to load pipeline schedulers : %s", err)
```
Comment (sguiheux) [inline engine/api/poller/cleaner.go]: pipeline schedulers => pollers?

File: engine/api/poller/cleaner.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+		return nil, err
+	}
+
+	deleted := []sdk.RepositoryPollerExecution{}
+	for _, s := range ps {
+		exs, err := LoadPastExecutions(tx, s.ApplicationID, s.PipelineID)
+		if err != nil {
+			log.Warning("poller.CleanerRun> Unable to load pipeline schedulers execution : %s", err)
```
Comment (sguiheux) [inline engine/api/poller/cleaner.go]: pipeline schedulers => poller

File: engine/api/poller/cleaner.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		}
+
+		nbToDelete := len(exs) - nbToKeep
+		nbDeleted := 0
+		for i := range exs {
+			if nbDeleted > nbToDelete {
+				break
+			}
+			if err := DeleteExecution(tx, &exs[i]); err != nil {
+				log.Critical("poller.CleanerRun> Unable to delete execution %d", exs[i].ID)
```
Comment (sguiheux) [inline engine/api/poller/cleaner.go]: log err too

File: engine/api/poller/dao.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+	"github.com/ovh/cds/engine/api/application"
+	"github.com/ovh/cds/engine/api/database"
+	"github.com/ovh/cds/engine/api/pipeline"
+	"github.com/ovh/cds/engine/log"
+	"github.com/ovh/cds/sdk"
+)
+
+//Insert insert or update a new poller in DB
+func Insert(db gorp.SqlExecutor, poller *sdk.RepositoryPoller) error {
```
Comment (sguiheux) [inline engine/api/poller/dao.go]: or update?

File: engine/api/poller/dao.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	p, err := unwrapPoller(db, dbPoller)
+	if err != nil {
+		return nil, err
+	}
+	return &p, nil
+}
+
+func postGet(db gorp.SqlExecutor, p *sdk.RepositoryPoller) error {
```
Comment (sguiheux) [inline engine/api/poller/dao.go]: in database package?

File: engine/api/poller/dao.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        FROM poller
+        WHERE application_id = $1
+		AND pipeline_id = $2
+    `
+	dbPoller := database.RepositoryPoller{}
+	if err := db.SelectOne(&dbPoller, query, applicationID, pipelineID); err != nil {
+		return nil, err
+	}
+
+	p, err := unwrapPoller(db, dbPoller)
```
Comment (sguiheux) [inline engine/api/poller/dao.go]: No need PostGET with select One

File: engine/api/poller/dao_execution.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
+package poller
+
+import (
+	"database/sql"
+
+	"github.com/go-gorp/gorp"
```
Comment (sguiheux) [inline engine/api/poller/dao_execution.go]: empty line after gorp

File: engine/api/poller/executer.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+		if err := executerProcess(tx, p, e); err != nil {
+			log.Critical("poller.ExecuterRun> Unable to process %v : %s", e, err)
+			tx.Rollback()
+			continue
+		}
+
+		//Commit
+		if err := tx.Commit(); err != nil {
+			log.Warning("poller.ExecuterRun> %s", err)
+			return nil, err
```
Comment (sguiheux) [inline engine/api/poller/executer.go]: rollback + continue

File: engine/api/poller/scheduler.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+//Scheduler is the goroutine which compute date of next execution for pipeline scheduler
+func Scheduler() {
+	for {
+		time.Sleep(2 * time.Second)
+		_, status, err := SchedulerRun()
+
+		if err != nil {
+			log.Critical("%s: %s", status, err)
```
Comment (sguiheux) [inline engine/api/poller/scheduler.go]: PollerScheduler> %s: %s  to know where err come from

File: engine/api/application/application.go (hunk 1 above)
Comment (fsamin) [inline engine/api/application/application.go]: nop nop nop

File: engine/api/poller/dao_execution.go (hunk 9 above)
Comment (fsamin) [inline engine/api/poller/dao_execution.go]: fixed

File: engine/api/database/gorp.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
+package database
+
+import (
+	"database/sql"
+	"log"
+
+	"github.com/go-gorp/gorp"
+
+	"github.com/ovh/cds/engine/api/database/gorpmapping"
+	"github.com/spf13/viper"
```
Comment (sguiheux) [inline engine/api/database/gorp.go]: wrong place

File: engine/api/hatchery/heartbeat.go [hunk 13]
```diff
@@ -3,7 +3,7 @@ package hatchery
 import (
 	"time"
 
-	"github.com/ovh/cds/engine/api/database"
+	"github.com/go-gorp/gorp"
 	"github.com/ovh/cds/engine/log"
```
Comment (sguiheux) [inline engine/api/hatchery/heartbeat.go]: need one empty line

File: engine/api/pipeline/gorp_model.go [hunk 14]
```diff
@@ -1,11 +1,19 @@
-package database
+package pipeline
 
 import (
 	"encoding/json"
 
 	"github.com/go-gorp/gorp"
+	"github.com/ovh/cds/engine/api/database/gorpmapping"
```
Comment (sguiheux) [inline engine/api/pipeline/gorp_model.go]: need one empty line above

File: engine/api/poller/cleaner.go [hunk 15]
```diff
@@ -1,30 +1,24 @@
 package poller
 
 import (
-	"fmt"
 	"time"
 
-	"github.com/ovh/cds/engine/api/database"
+	"github.com/go-gorp/gorp"
 	"github.com/ovh/cds/engine/log"
```
Comment (sguiheux) [inline engine/api/poller/cleaner.go]: need one empty line above

File: engine/api/poller/gorp_model.go [hunk 16]
```diff
@@ -0,0 +1,92 @@
+package poller
+
+import (
+	"encoding/json"
+
+	"github.com/go-gorp/gorp"
+	"github.com/ovh/cds/engine/api/application"
```
Comment (sguiheux) [inline engine/api/poller/gorp_model.go]: need empty line above

File: engine/api/poller/scheduler.go [hunk 17]
```diff
@@ -1,10 +1,9 @@
 package poller
 
 import (
-	"fmt"
 	"time"
 
-	"github.com/ovh/cds/engine/api/database"
+	"github.com/go-gorp/gorp"
 	"github.com/ovh/cds/engine/log"
```
Comment (sguiheux) [inline engine/api/poller/scheduler.go]: need one line above

File: engine/api/scheduler/gorp_model.go [hunk 18]
```diff
@@ -1,13 +1,19 @@
-package database
+package scheduler
 
 import (
 	"encoding/json"
 
 	"github.com/go-gorp/gorp"
-
+	"github.com/ovh/cds/engine/api/database/gorpmapping"
```
Comment (sguiheux) [inline engine/api/scheduler/gorp_model.go]: need one line above

File: engine/api/templateextension/gorp_model.go [hunk 19]
```diff
@@ -1,11 +1,16 @@
-package database
+package templateextension
 
 import (
 	"encoding/json"
 
 	"github.com/go-gorp/gorp"
+	"github.com/ovh/cds/engine/api/database/gorpmapping"
```
Comment (sguiheux) [inline engine/api/templateextension/gorp_model.go]: need one empty line above

File: engine/api/worker/gorp_model.go [hunk 20]
```diff
@@ -1,13 +1,16 @@
-package database
+package worker
 
 import (
 	"encoding/json"
 
 	"github.com/go-gorp/gorp"
-
+	"github.com/ovh/cds/engine/api/database/gorpmapping"
```
Comment (sguiheux) [inline engine/api/worker/gorp_model.go]: need one empty line above

File: engine/api/database/gorp.go (hunk 12 above)
Comment (fsamin) [inline engine/api/database/gorp.go]: fixed

**Pull request merged**